Improve partitioned output operator performance for strings

lukasz-stec opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description
Improve performance of `SlicePositionsAppender`.
The improvement comes from not allocating `Slice` per position + using `System.arrayCopy` directly instead of `Slice.getBytes`

> Is this change a fix, improvement, new feature, refactoring, or other?

performance improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine (partitioned exchange) and SPI extension

> How would you describe this change to a non-technical end user or system administrator?

Improve performance of remote partitioned exchange

### Benchamrks
#### jmh
about 30% improvement
```
Before
Benchmark                                   (channelCount)  (enableCompression)  (nullRate)  (partitionCount)  (positionCount)   (type)  Mode  Cnt     Score     Error  Units
BenchmarkPartitionedOutputOperator.addPage               1                false           0                16             8192  VARCHAR  avgt   20  2137.367 ± 166.834  ms/op
BenchmarkPartitionedOutputOperator.addPage               1                false         0.2                16             8192  VARCHAR  avgt   20  1660.550 ±  24.274  ms/op

After
BenchmarkPartitionedOutputOperator.addPage               1                false           0                16             8192  VARCHAR  avgt   20  1399.476 ± 181.301  ms/op
BenchmarkPartitionedOutputOperator.addPage               1                false         0.2                16             8192  VARCHAR  avgt   20  1194.077 ± 160.496  ms/op
```

#### tpch/tpcds
![image](https://user-images.githubusercontent.com/8080198/173786440-1a0ba4f7-9266-4f0b-b15e-73369f2bc682.png)
Overall there is a small improvement. This is expected as `PartitionedOutputOperator` is only around 5% of total CPU time and this pr only improves varchars.
When looking at the `PartitionedOutputOperator` stats only, the improvement is ~10%.
```
Before
Total: 980917.6656576941
PartitionedOutputOperator: 43522.99081258998, 0.044369667645253706%
After
Total: 974919.225831081
PartitionedOutputOperator: 38515.177167889924, 0.03950601870124904%
```
[poo-slice-orc-part-sf1K.pdf](https://github.com/trinodb/trino/files/8907614/poo-slice-orc-part-sf1K.pdf)

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Improve performance of remote partitioned exchange on string data types. ({issue}`12798`)
```

raunaqmorarka commented on core/trino-spi/src/main/java/io/trino/spi/block/AbstractVariableWidthBlock.java:
```diff
@@ -120,6 +120,23 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
         blockBuilder.writeBytes(getRawSlice(position), getPositionOffset(position) + offset, length);
     }
 
+    public void writeBytesTo(int[] positions, int positionsCount, BatchByteWriter byteWriter)
```
Could this API be
`void copyPositions(int[] positions, int offset, int length, byte[] destination, int destinationOffset)` ?
You can also look into using similar logic as `copyPositions`, it would just copy into provided bytes array from given target offset instead of a Slice.

raunaqmorarka replied:
Aren't you already doing an extra loop in `process...Positions` to figure out size of destination array ? You can potentially just do that before this method to size destination bytes array correctly. Or you can send in an output buffer class instead of raw bytes array which will apply resizing logic of ensureBytesCapacity when needed.
Can we make the API `void copyPositions(int[] positions, int srcIndex, int length, DynamicSliceOutput output, int[] positionOffsets, boolean[] isNull, int destIndex)` ? This might enable us to extract what we need without requiring multiple passes.

lukasz-stec replied:
> Aren't you already doing an extra loop in process...Positions to figure out size of destination array ? You can potentially just do that before this method to size destination bytes array correctly.

Yes but this loop does not calculate lengths, it uses it to update offsets and other appender state.

> Can we make the API void copyPositions(int[] positions, int srcIndex, int length, DynamicSliceOutput output, int[] positionOffsets, boolean[] isNull, int destIndex) ? This might enable us to extract what we need without requiring multiple passes.

so this method would also update `positionOffsets` and `isNull`? why is that better than the original API?

raunaqmorarka replied:
> so this method would also update positionOffsets and isNull?

Yes, it would basically make this one method make all the necessary updates.
If it works, it would be better as it would avoid introducing `BatchByteWriter`. It could avoid the need for multiple passes, one to find non-nulls and another over the non-nulls to find offset and length. It's also easier to read as it would consolidate most of the logic in one place instead of having to hop between between caller of this method and the logic inside this method.

raunaqmorarka commented on core/trino-spi/src/main/java/io/trino/spi/block/AbstractVariableWidthBlock.java:
```diff
@@ -120,6 +120,23 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
+        int[] offsets = new int[positionsCount];
+        for (int i = 0; i < positionsCount; i++) {
+            int position = positions[i];
+            if (!isNull(position)) {
```
There is no need for `isNull` check. Slice length will be 0 and position offset will be same as the last position's offset.

raunaqmorarka commented on core/trino-spi/src/main/java/io/trino/spi/block/AbstractVariableWidthBlock.java:
```diff
@@ -120,6 +120,23 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
+        for (int i = 0; i < positionsCount; i++) {
+            int position = positions[i];
+            if (!isNull(position)) {
+                int length = getSliceLength(position);
+                int offset = getPositionOffset(position);
```
We're looking into offset array 3 times here, you can try something like below
```
        int[] offsets = new int[positionsCount];
        int finalLength = 0;
        for (int i = 0; i < positionsCount; i++) {
            int position = positions[i];
            finalLength += getSliceLength(position);
            offsets[i + 1] = finalLength;
        }
```

lukasz-stec replied:
Positions are not necessarily consecutive so the lookups for the given and the subsequent position are necessary.
I expect the compiler to merge two lookups for the current positions into one.
We could check if the positions are consecutive and avoid one lookup at the cost of a branch but I'm not sure it would be beneficial in all cases

raunaqmorarka replied:
Okay non-consecutive positions rules that out.

lukasz-stec reviewed (commented):
comments answered.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -82,38 +83,20 @@ public void append(IntArrayList positions, Block block)
-        int[] positionArray = positions.elements();
-        int newByteCount = 0;
-        int[] lengths = new int[positions.size()];
 
```
We can make use of protected APIs by writing a method in BlockUtil and exposing that. It's not ideal but it might be better than adding the proposed new SPI methods from current PR.

lukasz-stec commented:
I rebased on the master (there were some changes to the `SlicePositionsAppender`) and as discussed offline I reverted the `BatchByteWriter` change in favor of exposing raw `Slice` and offsets from the `VariableWidthBlock`.

lukasz-stec commented:
I also updated the jmh results as the absolute number changed significantly after rebasing on the master and running the benchmarks on java 11.0.15 instead of 11.0.11. The relative improvements are similar (slightly better).

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
-            int position = positions[i];
-            if (!block.isNull(position)) {
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
```
Since you are actively avoiding use of `Slice#getBytes(int index, byte[] destination, int destinationIndex, int length)` here, could you check if the benefit is coming from simply skipping the range checks done inside that method or is `System.arraycopy` faster than `unsafe.copyMemory` ?
If it is the latter, you could just improve `Slice#getBytes` implementation to use `System.arraycopy` for byte array case (it would also be interesting to know why `System.arraycopy` is faster than `unsafe.copyMemory`). If it is the former, you could look into providing an unchecked version of `Slice#getBytes` in Slice.

raunaqmorarka replied:
I attempted to move `System.arraycopy` to `Slice.getBytes`, but that was slower than the current code in JMH benchmarks. So I suspect at least some of the improvement here is due to fewer positions checks and/or manual inlining. For the time being I've applied some fixups but kept same approach.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
         for (int i = 0; i < count; i++) {
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
-            offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
+            this.offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
```
this.offsets -> offsets

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
+        }
         for (int i = 0; i < count; i++) {
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
-            offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
+            this.offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
```
nit: Could be simplified to something like
```
int currentStartOffset = startOffset + length;
for (int i = 0; i < count; i++) {
    offsets[positionCount + i + 1] = currentStartOffset;
    currentStartOffset += length;
}
```

lukasz-stec replied:
I missed this comment.
applied now.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
 
-        Slice slice = block.getSlice(position, 0, length);
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
+            for (int i = 0; i < count; i++) {
+                System.arraycopy(base, sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                rawSlice.getBytes(sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
```
An alternative way here is to copy only once into the byte array from Slice using this method, then use `System.arrayCopy` to copy bytes within the byte array with doubling size at each step, so you get only log_2(count) calls to `System.arrayCopy` overall.
Not sure if it'll be faster though.

lukasz-stec replied:
This is a good idea.
First of all, it allows us to not rely on Slice being backed by a byte array.
Second, I did a quick benchmark and this method is ridiculously fast on small to medium chunk sizes.
I suspect it's not gonna be that used in practice but I included this in this PR given results below.
```
Benchmark                               (length)  Mode  Cnt      Score   Error  Units
BenchmarkDuplicateBytes.arrayCopy              2  avgt   20      4.101 ±   0.082 ns/op
BenchmarkDuplicateBytes.arrayCopy             16  avgt   20      5.088 ±   0.055 ns/op
BenchmarkDuplicateBytes.arrayCopy             35  avgt   20      4.389 ±   0.175 ns/op
BenchmarkDuplicateBytes.arrayCopy             32  avgt   20      4.333 ±   0.049 ns/op
BenchmarkDuplicateBytes.arrayCopy             64  avgt   20      4.900 ±   0.050 ns/op
BenchmarkDuplicateBytes.arrayCopy             80  avgt   20      5.609 ±   0.117 ns/op
BenchmarkDuplicateBytes.arrayCopy            128  avgt   20      5.318 ±   0.069 ns/op
BenchmarkDuplicateBytes.arrayCopy            200  avgt   20      6.087 ±   0.098 ns/op
BenchmarkDuplicateBytes.arrayCopy            256  avgt   20      6.971 ±   0.111 ns/op
BenchmarkDuplicateBytes.arrayCopy            512  avgt   20     12.175 ±   0.324 ns/op
BenchmarkDuplicateBytes.arrayCopy           1024  avgt   20     22.759 ±   0.300 ns/op
BenchmarkDuplicateBytes.arrayCopy           2048  avgt   20     55.769 ±   1.213 ns/op
BenchmarkDuplicateBytes.arrayCopy         100000  avgt   20   4931.827 ±  276.57 ns/op
BenchmarkDuplicateBytes.arrayCopy        1000000  avgt   20  70111.621 ± 922.907 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x         2  avgt   20      0.230 ±   0.005 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        16  avgt   20      0.352 ±   0.015 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        35  avgt   20      0.533 ±   0.052 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        32  avgt   20      0.498 ±   0.007 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        64  avgt   20      0.881 ±   0.013 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        80  avgt   20      1.100 ±   0.015 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       128  avgt   20      1.518 ±   0.228 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       200  avgt   20      3.522 ±   0.069 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       256  avgt   20      4.472 ±   0.154 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       512  avgt   20     10.083 ±   0.206 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x      1024  avgt   20     21.741 ±   0.349 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x      2048  avgt   20     57.329 ±   0.854 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x    100000  avgt   20   5243.664 ± 188.963 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x   1000000  avgt   20 109400.739 ± 992.325 ns/op
```

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -82,38 +83,41 @@ public void append(IntArrayList positions, Block block)
                 int position = positionArray[i];
-                int length = block.getSliceLength(position);
+                int length = variableWidthBlock.getSliceLength(position);
+                int offset = variableWidthBlock.getPositionOffset(position);
```
nit: inline `offset`

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -82,38 +83,41 @@ public void append(IntArrayList positions, Block block)
-                    hasNonNullValue = true;
-                }
+                int length = variableWidthBlock.getSliceLength(position);
+                int offset = variableWidthBlock.getPositionOffset(position);
```
nit: inline `offset`

raunaqmorarka reviewed (commented):
minor comments, looks good overall

raunaqmorarka commented:
https://github.com/trinodb/trino/issues/13107

skrzypo987 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
                 int length = lengths[i];
-                Slice slice = block.getSlice(position, 0, length);
-                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+                if (length > 0) {
```
This check is totally done withing `System.arraycopy`

skrzypo987 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,39 +170,56 @@ public long getSizeInBytes()
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
             offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
         }
-
```
Irrelevant blank line deletion

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java:
```diff
@@ -218,7 +218,7 @@ public Block copyPositions(int[] positions, int offset, int length)
 
     @Override
-    protected Slice getRawSlice(int position)
+    public Slice getRawSlice(int position)
```
Maintainers actually agreed on that?

skrzypo987 reviewed (commented):
Looks ok % comment about making `getRawSlice` method public

sopel39 commented:
https://github.com/trinodb/trino/issues/13267 is most likely a prerequisite

lukasz-stec commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,16 +170,20 @@ public long getSizeInBytes()
     }
 
-    private void copyBytes(Block block, int[] lengths, int[] positions, int count, int[] targetOffsets, int targetOffsetsIndex, int newByteCount)
+    private void copyBytes(Slice rawSlice, int[] lengths, int[] sourceOffsets, int count, int newByteCount)
```
As I mentioned in the description, part of the gain is from avoiding `getSlice` (~10% overall, 1/3 of the gain) and the other part is from using `System.arraycopy` instead of `Slice.getBytes`.
Theoretically allocating `Slice` in `getSlice` could be eliminated by inlining and escape analysis but it does not happen in practice, at least not in the jmh (the call is inlined but allocating is not eliminated).

lukasz-stec commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -168,16 +170,20 @@ public long getSizeInBytes()
-                int length = lengths[i];
-                Slice slice = block.getSlice(position, 0, length);
-                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+        if (rawSlice.hasByteArray()) {
```
yes, it gives ~20% improvement.

lukasz-stec reviewed (commented):
a few comments answered

lukasz-stec reviewed (commented):
comments addressed + added `if (block instanceof VariableWidthBlock) {` + generic path + rebaed on the master.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -80,38 +80,43 @@ public void append(IntArrayList positions, Block block)
+                    lengths[i] = length;
+                    sourceOffsets[i] = variableWidthBlock.getRawSliceOffset(position);
+                    newByteCount += length;
+                    boolean isNull = block.isNull(position);
```
just a thought:
We can add Slice#getNulls(boolean[] to, int offset, int length) or sth like that to copy the entire portion of isNull array and get rid of bound checks this way. 
But this is clearly out of scope here. Just wanted to write it down somewhere

lukasz-stec replied:
we would have to then pass `positions` to the `block.getNulls` (positions are not consecutive).
I did experiments with this idea but I don't see any improvement

sopel39 commented on core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java:
```diff
@@ -85,6 +86,23 @@ public VariableWidthBlock(int positionCount, Slice slice, int[] offsets, Optiona
+    /**
+     * Gets the raw {@link Slice} that keeps the actual data bytes.
+     */
+    public Slice getRawSlice()
```
move these methods to `AbstractVariableWidthBlock`. It will be more generic then

lukasz-stec replied:
talked offline, AbstractVariableWidthBlock may not support one single Slice. I'm leaving this here only.

sopel39 commented on core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java:
```diff
@@ -85,6 +86,23 @@ public VariableWidthBlock(int positionCount, Slice slice, int[] offsets, Optiona
+     */
+    public int getRawSliceOffset(int position)
+    {
+        checkValidPosition(position, getPositionCount());
```
most of the methods use `checkReadablePosition`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -80,38 +80,43 @@ public void append(IntArrayList positions, Block block)
-                    offsets[positionCount + i + 1] = offsets[positionCount + i];
-                    valueIsNull[positionCount + i] = true;
-                    hasNullValue = true;
+        if (block instanceof VariableWidthBlock) {
```
`if (block instanceof VariableWidthBlock) {`
->
`if (block instanceof AbstractVariableWidthBlock) {`

lukasz-stec replied:
there is a possible performance impact if w have more than 1 type here. Also, with `VariableWidthBlock`, just CHA can provide optimal compilation here, even with C1 as there is only one type. With `AbstractVariableWidthBlock` there are always 2 (builder) so CHA cannot fully help here so jit has to rely on the call site profile

lukasz-stec replied:
talked offline, `AbstractVariableWidthBlock` may not suppoort one single Slice. I'm leaving `VariableWidthBlock`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -80,38 +80,43 @@ public void append(IntArrayList positions, Block block)
+                    newByteCount += length;
+                    boolean isNull = block.isNull(position);
+                    valueIsNull[positionCount + i] = isNull;
+                    offsets[positionCount + i + 1] = offsets[positionCount + i] + length;
```
Is there a guarantee that `length` is 0 for `null` position?

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -80,38 +80,43 @@ public void append(IntArrayList positions, Block block)
+                    valueIsNull[positionCount + i] = isNull;
+                    offsets[positionCount + i + 1] = offsets[positionCount + i] + length;
+                    hasNullValue |= isNull;
+                    hasNonNullValue |= !isNull;
```
I wonder if it would be faster if these fields were local to the loop and class-level fields were updated only after the loop

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -131,8 +136,18 @@ public void appendRle(RunLengthEncodedBlock block)
             updateSize(rlePositionCount, 0);
         }
         else {
+            if (block.getValue() instanceof VariableWidthBlock) {
```
I don't think you have to specialize here. You can just use block generic `io.trino.spi.block.Block#getSlice`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -184,25 +203,58 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
+            }
+            else {
+                int length = block.getSliceLength(position);
+                ensureBytesCapacity(offsets[positionCount] + length);
```
`offsets[positionCount]` -> `getCurrentOffset`.

but maybe `ensureBytesCapacity` should only be used with delta rather than absolute value, like `ensureExtraBytesCapacity`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,6 +265,13 @@ public void testSliceRle()
     }
 
+    @Test
+    public void testSliceAppenderInputNotByteArray()
```
do same for RLE?

sopel39 replied:
Move to `TestSlicePositionsAppender`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
         }
     }
+
+    public static Block createTesVarcharBlock(String... values)
```
`createTesVarcharBlock` -> `createTestVarcharBlock`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
+    }
+
+    // Fake varchar type to test handling not VariableWidthBlock blocks in SlicePositionAppender
+    static class TestVarcharType
```
Why do we need new varchar type? You can use standard varchar with special block type

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
+        }
+    }
+
+    private static class TestVariableWidthBlockBuilder
```
Do we need to have a special builder? Maybe you could just add some kind of adaptation that would convert `VariableWidthBlock` to `TestVariableWidthBlock`?

lukasz-stec replied:
I added adaptation instead of Type and Builder

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
+            }
+            this.positionCount = positionCount;
+
+            if (slice == null) {
```
you can use `requireNotNull` here

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
+
+        private TestVariableWidthBlock(int arrayOffset, int positionCount, Slice slice, int[] offsets, boolean[] valueIsNull)
+        {
+            if (arrayOffset < 0) {
```
use `checkArgument`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -463,4 +509,290 @@ public void appendTo(PositionsAppender positionsAppender)
+                throw new IllegalArgumentException("arrayOffset is negative");
+            }
+            this.arrayOffset = arrayOffset;
+            if (positionCount < 0) {
```
use `checkArgument`

lukasz-stec reviewed (commented):
comments addressed

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -184,25 +195,73 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
 
-        Slice slice = block.getSlice(position, 0, length);
+        if (length > 0) {
+            duplicateBytes(slice, sourceOffset, bytes, startOffset, length, count);
```
`if (length > 0) {` check is redundant (there is check already in `duplicateBytes`)

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -184,25 +195,73 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
-            offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
+            offsets[positionCount + i + 1] = currentStartOffset;
+            currentStartOffset += length;
```
nit: isn't multiplication faster?

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +252,42 @@ public void testSliceRle()
+    }
+
+    @Test
+    public void testEmptySliceRLe()
```
`testEmptySliceRLe` -> `testEmptySliceRle`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +252,42 @@ public void testSliceRle()
+    @Test
+    public void testEmptySliceRLe()
+    {
+        testAppend(VARCHAR, ImmutableList.of(
```
move to `TestSlicePositionAppender`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +252,42 @@ public void testSliceRle()
+        assertBlockIsValid(actual, sizeInBytes, VARCHAR, ImmutableList.of(new BlockView(rleBlock, allPositions(rleBlock.getPositionCount()))));
+    }
+
+    @Test
```
isn't this already covered by another test? Add comment

lukasz-stec replied:
I don't think so, these tests usually use randomly generated values and an empty string is a specific value that I want to test.
comment added.

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +252,42 @@ public void testSliceRle()
+    }
+
+    @Test
+    public void testSliceAppenderInputRleNotByteArray()
```
Isn't this already covered by adaptation?

sopel39 replied:
tests with block adaptation

lukasz-stec replied:
Ah, this new one. So no, it's not covered by normal tests, even with adaptation. Like I mentioned [here](https://github.com/trinodb/trino/pull/12798#discussion_r959540741) I need to access `SlicePositionsAppender` directly because `UnnestingPositionsAppender` copies RLE value to byte array always.
That said, in the current implementation I don't rely on the byte array in the RLE block anymore so I could drop this test

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -76,66 +86,66 @@
 
     @Test(dataProvider = "types")
-    public void testMixedBlockTypes(Type type)
+    public void testMixedBlockTypes(Type type, Function<Block, Block> blockAdaptation)
```
Introduce composite `TestType` that composes of `Type type` and `Function<Block, Block> blockAdaptation`
Then you wouldn't have to pass `blockAdaptation` around and it will be more explicit what is tested, e.g:
```
VARCHAR,
VARCHAR_WITH_TEST_BLOCK
```
`TestType` could have default constructor so that you wouldn't have to use `identity` arg by default

lukasz-stec replied:
done

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java:
```diff
@@ -184,25 +194,74 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
+     * Copy {@code length} bytes from {@code slice}, starting at offset {@code sourceOffset} to {@code count} consecutive positions in the {@link #bytes} array.
+     */
+    @VisibleForTesting
+    static void duplicateBytes(Slice slice, int sourceOffset, byte[] bytes, int startOffset, int length, int count)
```
drop `sourceOffset` and `length`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +253,34 @@ public void testSliceRle()
         }
     }
 
+    @Test
```
rename test name, if it's not byte array, then what it is?

Why the test is important (please add short comment)

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -241,14 +253,34 @@ public void testSliceRle()
+                input(new VariableWidthBlock(3, Slices.wrappedLongArray(257, 2), new int[] {0, 1, Long.BYTES, 2 * Long.BYTES}, Optional.empty()), 0, 2)));
+    }
+
+    @Test
```
rename, `NotByteArray` -> `WithSomeType`

lukasz-stec replied:
I dropped this test. it's no longer needed.

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -260,44 +292,15 @@ public void testRowWithNestedFields()
-                        {createTimestampType(9)},
-                        {anonymousRow(BIGINT, VARCHAR)}
-                };
+        return Arrays.stream(TestType.values()).filter(TestType::isSupportsNullRle).map(type -> new Object[] {type}).toArray(Object[][]::new);
```
split into multiple lines:
```
.filter
.map
toArray
```

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -260,44 +292,15 @@ public void testRowWithNestedFields()
-                        {createTimestampType(9)},
-                        {anonymousRow(BIGINT, VARCHAR)}
-                };
+        return Arrays.stream(TestType.values()).map(type -> new Object[] {type}).toArray(Object[][]::new);
```
split into mulitple lines

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -437,6 +445,61 @@ private Block buildBlock(Type type, List<BlockView> inputs, BlockBuilderStatus b
         return blockBuilder.build();
     }
 
+    enum TestType
```
private

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -437,6 +445,61 @@ private Block buildBlock(Type type, List<BlockView> inputs, BlockBuilderStatus b
+
+        private final boolean supportsNullRle;
+
+        TestType(Type type)
```
public (and below)

lukasz-stec replied:
enum constructors have to be private

sopel39 replied:
Make it private then (not package private)

lukasz-stec replied:
it is private. one does not simply change the enum constructor visibility :-)

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java:
```diff
@@ -437,6 +445,61 @@ private Block buildBlock(Type type, List<BlockView> inputs, BlockBuilderStatus b
+            this(type, Function.identity(), supportsNullRle);
+        }
+
+        TestType(Type type, Function<Block, Block> blockAdaptation, boolean supportsNullRle)
```
everything should support `supportsNullRle`

lukasz-stec replied:
`ArrayBlockBuilder` does not support it

sopel39 replied:
> ArrayBlockBuilder does not support it

Does it make sense to add support there? It seems odd that we have to work around that particular type

lukasz-stec replied:
ok, created https://github.com/trinodb/trino/pull/13973

lukasz-stec replied:
moved changes from  https://github.com/trinodb/trino/pull/13973 to the first commit here + adjusted the tests

lukasz-stec reviewed (commented):
ca

sopel39 commented on plugin/trino-thrift-api/src/main/java/io/trino/plugin/thrift/api/datatypes/TrinoThriftBigintArray.java:
```diff
@@ -144,12 +145,18 @@ public String toString()
+        if (block instanceof RunLengthEncodedBlock && block.isNull(0)) {
+            boolean[] nulls = new boolean[positions];
+            Arrays.fill(nulls, true);
+            return bigintArrayData(new TrinoThriftBigintArray(nulls, null, null));
```
Can you add a test?

This would fail
```
checkArgument(totalSize(nulls, sizes) == numberOfValues(values), "total number of values doesn't match expected size");
```
in `TrinoThriftBigintArray`

lukasz-stec replied:
> Can you add a test?

`TestReadWrite.testSingleRowPageReadWrite` already tests null RLE 



> This would fail
> 
> ```
> checkArgument(totalSize(nulls, sizes) == numberOfValues(values), "total number of values doesn't match expected size");
> ```
> 
> in `TrinoThriftBigintArray`

WDYM? can this fail when `nulls` is all true and `sizes` and `values` are both null?

sopel39 replied:
Yes. It would fail. Please inspect code. I think test is lacking

lukasz-stec replied:
> Yes. It would fail. Please inspect code (mentioned assertion). I think test is lacking


Could you be more specific? I think my code inspection skills are lacking..
My _inspection_ is this: `numberOfValues(values)` always returns 0 for `null` `values`.
`totalSize` also returns 0 if all nulls are set to true because this condition is always false `if (nulls == null || !nulls[i]) {`

sopel39 replied:
Ah. You are right

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,146 @@
+    @Test
+    public void testAppendEmptySliceRle()
+    {
+        // test SlicePositionAppender.appendRLe with empty value (sString/Slice with length 0)
```
`appendRLe` -> `appendRle`

`sString/Slice`??

lukasz-stec replied:
it was supposed to be `String/Slice` but lets make it just Slice

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Test
+    public void testDuplicateMultipleBytesPowerOFTwoNumberOfTimes()
```
`PowerOFTwo` -> `PowerOfTwo`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,146 @@
+        assertBlockEquals(VARCHAR, actualBlock, rleBlock);
+    }
+
+    // test append with VariableWidthBlock using Slice not supported by byte array
```
`using Slice not supported by byte array` -> `using Slice not backed by byte array`

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,146 @@
+
+        Block expected = new VariableWidthBlock(
+                2,
+                Slices.wrappedBuffer(ByteBuffer.allocate(9).order(nativeOrder()).put((byte) 1).putLong(2).flip()),
```
could you just put actual values there?
It's super hard to figure if this is a correct result or not

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,146 @@
+public class TestSlicePositionsAppender
+{
+    @Test
+    public void testAppendEmptySliceRle()
```
why this can't be part of `TestPositionsAppender`, e.g: `testAppendEmptyRle` (for given type)

lukasz-stec replied:
it was part of the `TestPositionsAppender` but not per type. Empty string/slice is only valid for `VARCHAR`.

sopel39 reviewed (commented):
lgtm % comments

lukasz-stec reviewed (commented):
ca

sopel39 reviewed (commented):
Added comment

sopel39 commented on core/trino-main/src/test/java/io/trino/operator/output/TestSlicePositionsAppender.java:
```diff
@@ -0,0 +1,144 @@
+    // test append with VariableWidthBlock using Slice not backed by byte array
+    // to test special handling in SlicePositionsAppender.copyBytes
+    @Test
+    public void testAppendSliceNotSupportedByByteArray()
```
`testAppendSliceNotSupportedByByteArray` -> `testAppendSliceNotBackedByByteArray`

lukasz-stec replied:
done

sopel39 reviewed (approved):
% comment

sopel39 commented:
conflict, please rebase

lukasz-stec commented:
> conflict, please rebase

rebased, checks in progress
